File: LovePlay-RAC/LovePlay/Class/Zone/View/DiscussDetailPostCell.m
//
//  DiscussDetailPostCell.m
//  LovePlay
//
//  Created by weiying on 2017/2/7.
//  Copyright © 2017年 Yuns. All rights reserved.
//

#import "DiscussDetailPostCell.h"
#import "DiscussDetailModel.h"

@interface DiscussDetailPostCell ()
//UI
@property (nonatomic, strong) UIImageView *imageNode;
@property (nonatomic, strong) UILabel *nameTextNode;
@property (nonatomic, strong) UILabel *timeTextNode;
@property (nonatomic, strong) UILabel *floorTextNode;
@property (nonatomic, strong) UILabel *contentTextNode;
@property (nonatomic, strong) UIView *underLineNode;
//Data
@property (nonatomic, strong) DiscuzPost *post;
@property (nonatomic, assign) NSInteger floor;
@end

@implementation DiscussDetailPostCell

+ (instancetype)cellWithTableView:(UITableView *)tableView
{
    static NSString *ID = @"DiscussDetailPostCell";
    DiscussDetailPostCell *cell = [tableView dequeueReusableCellWithIdentifier:ID];
    if (cell == nil) {
        cell = [[DiscussDetailPostCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:ID];
    }
    return cell;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        [self addSubnodes];
        [self mas_subViews];
    }
    return self;
}

#pragma mark - private
- (void)addSubnodes
{
    [self.contentView addSubview:self.imageNode];
    
    [self.contentView addSubview:self.nameTextNode];
    
    [self.contentView addSubview:self.timeTextNode];
    
    [self.contentView addSubview:self.floorTextNode];
    
    [self.contentView addSubview:self.contentTextNode];
    
    [self.contentView addSubview:self.underLineNode];
}

- (void)mas_subViews
{
    [_imageNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.height.mas_equalTo(30);
        make.top.left.equalTo(self.contentView).offset(10);
    }];
    
    [_floorTextNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self.contentView).offset(-10);
        make.centerY.equalTo(_imageNode);
    }];
    
    [_nameTextNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(_imageNode.mas_right).offset(10);
        make.centerY.equalTo(_imageNode);
    }];
    
    //放在name和floor之间，设置左右位置，并控制优先级
    [_timeTextNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(_nameTextNode.mas_right).offset(10).priorityMedium();
        make.right.equalTo(_floorTextNode.mas_left).offset(-10).priorityLow();
        make.centerY.equalTo(_imageNode);
    }];
    
    [_contentTextNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_imageNode.mas_bottom).offset(10);
        make.left.equalTo(self.contentView).offset(10);
        make.right.equalTo(self.contentView).offset(-10);
    }];
    
    [_underLineNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(0.5);
        make.top.equalTo(_contentTextNode.mas_bottom).offset(10);
        make.left.bottom.right.equalTo(self.contentView);
    }];
}

#pragma mark - public
- (void)setupPost:(DiscuzPost *)post floor:(NSInteger)floor
{
    _post = post;
    _floor = floor;
    _nameTextNode.text = post.author;
    _timeTextNode.text = [post.dateline stringByReplacingOccurrencesOfString:@"&nbsp;" withString:@""];
    _floorTextNode.text = [NSString stringWithFormat:@"%@#", @(floor).stringValue];
    _contentTextNode.text = post.message;
}

#pragma mark - setter / getter
- (UIImageView *)imageNode
{
    if (!_imageNode) {
        UIImageView *imageNode = [[UIImageView alloc] init];
        imageNode.image = [UIImage imageNamed:@"defult_pho"];
        _imageNode = imageNode;
    }
    return _imageNode;
}

- (UILabel *)nameTextNode
{
    if (!_nameTextNode) {
        UILabel *nameTextNode = [[UILabel alloc] init];
        nameTextNode.font = [UIFont systemFontOfSize:14];
        nameTextNode.textColor = RGB(186, 177, 161);
        _nameTextNode = nameTextNode;
    }
    return _nameTextNode;
}

- (UILabel *)timeTextNode
{
    if (!_timeTextNode) {
        UILabel *timeTextNode = [[UILabel alloc] init];
        timeTextNode.font = [UIFont systemFontOfSize:12];
        timeTextNode.textColor = RGB(163, 163, 163);
        _timeTextNode = timeTextNode;
    }
    return _timeTextNode;
}

- (UILabel *)floorTextNode
{
    if (!_floorTextNode) {
        UILabel *floorTextNode = [[UILabel alloc] init];
        floorTextNode.font = [UIFont systemFontOfSize:12];
        floorTextNode.textColor = RGB(163, 163, 163);
        floorTextNode.textAlignment = NSTextAlignmentRight;
        _floorTextNode = floorTextNode;
    }
    return _floorTextNode;
}

- (UILabel *)contentTextNode
{
    if (!_contentTextNode) {
        UILabel *contentTextNode = [[UILabel alloc] init];
        contentTextNode.numberOfLines = 0;
        contentTextNode.font = [UIFont systemFontOfSize:14];
        contentTextNode.textColor = RGB(50, 50, 50);
        _contentTextNode = contentTextNode;
    }
    return _contentTextNode;
}

- (UIView *)underLineNode
{
    if (!_underLineNode) {
        UIView *underLineNode = [[UIView alloc] init];
        underLineNode.backgroundColor = RGB(222, 222, 222);
        _underLineNode = underLineNode;
    }
    return _underLineNode;
}
@end

File: LovePlay-Swift3/LovePlay/Class/Zone/View/DiscussDetailPostCell.swift
//
//  DiscussDetailPostCell.swift
//  LovePlay
//
//  Created by Yuns on 2017/6/11.
//  Copyright © 2017年 yuns. All rights reserved.
//

import UIKit

class DiscussDetailPostCell: UITableViewCell {

    var post : DiscussDetailPost?
    var floor : Int?
    
    class func cellWithTableView(tableView : UITableView) -> DiscussDetailPostCell {
        let ID = NSStringFromClass(self)
        var cell : DiscussDetailPostCell? = tableView.dequeueReusableCell(withIdentifier: ID) as? DiscussDetailPostCell
        if cell == nil {
            cell = DiscussDetailPostCell(style: .default, reuseIdentifier: ID)
        }
        return cell!
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.selectionStyle = .none
        self.addSubViews()
        self.snp_subViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    // MARK: - private
    private func addSubViews() {
        self.contentView.addSubview(self.imgView)
        self.contentView.addSubview(self.nameTextLabel)
        self.contentView.addSubview(self.timeTextLabel)
        self.contentView.addSubview(self.floorTextLabel)
        self.contentView.addSubview(self.contentTextLabel)
        self.contentView.addSubview(self.underLineView)
    }
    
    private func snp_subViews() {
        self.imgView.snp.makeConstraints { (make) in
            make.width.height.equalTo(30)
            make.top.left.equalToSuperview().offset(10)
        }
        
        self.nameTextLabel.snp.makeConstraints { (make) in
            make.left.equalTo(self.imgView.snp.right).offset(10)
            make.centerY.equalTo(self.imgView)
        }
        
        self.timeTextLabel.snp.makeConstraints { (make) in
            make.left.equalTo(self.nameTextLabel.snp.right).offset(10).priority(.medium)
            make.right.equalTo(self.floorTextLabel.snp.left).offset(-10).priority(.low)
            make.centerY.equalTo(self.imgView)
        }
        
        self.floorTextLabel.snp.makeConstraints { (make) in
            make.right.equalToSuperview().offset(-10)
            make.centerY.equalTo(self.imgView)
        }
        
        self.contentTextLabel.snp.makeConstraints { (make) in
            make.top.equalTo(self.imgView.snp.bottom).offset(10)
            make.left.equalToSuperview().offset(10)
            make.right.equalToSuperview().offset(-10)
        }
        
        self.underLineView.snp.makeConstraints { (make) in
            make.height.equalTo(0.5)
            make.top.equalTo(self.contentTextLabel.snp.bottom).offset(10)
            make.left.bottom.right.equalToSuperview()
        }
    }
    
    // MARK: - publick
    public func setupDiscussDetailComment(post : DiscussDetailPost, floor : Int) {
        self.post = post
        self.floor = floor
        
        self.nameTextLabel.text = post.author
        self.timeTextLabel.text = post.dateline?.replacingOccurrences(of: "&nbsp;", with: "")
        self.floorTextLabel.text = "\(floor)#"
        self.contentTextLabel.text = post.message
    }
    
    // MARK: - setter / getter
    lazy var imgView : UIImageView = {
        let imgView : UIImageView = UIImageView()
        imgView.image = UIImage(named: "defult_pho")
        return imgView
    }()
    
    lazy var nameTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.font = UIFont.systemFont(ofSize: 14)
        label.textColor = RGB(r: 186, g: 177, b: 161)
        return label
    }()
    
    lazy var timeTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.textColor = RGB(r: 163, g: 163, b: 163)
        return label
    }()
    
    lazy var floorTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.textColor = RGB(r: 163, g: 163, b: 163)
        return label
    }()
    
    lazy var contentTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.numberOfLines = 0
        label.font = UIFont.systemFont(ofSize: 14)
        label.textColor = RGB(r: 50, g: 50, b: 50)
        return label
    }()
    
    lazy var underLineView : UIView = {
        let lineView : UIView = UIView()
        lineView.backgroundColor = RGB(r: 222, g: 222, b: 222)
        return lineView
    }()

}
